File: src/security/sanitization.py
"""
Input/output sanitization — strip dangerous content before LLM or logging.
"""

import re
import unicodedata
from typing import Any, Dict, List, Union

from src.security.redaction import redact_pii

# Control chars except common whitespace used in chat
_CONTROL_CHAR_RE = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]")
_NULL_BYTE_RE = re.compile(r"\x00")

_SCRIPT_TAG_RE = re.compile(
    r"<script[^>]*>.*?</script>",
    re.IGNORECASE | re.DOTALL,
)
_IFRAME_TAG_RE = re.compile(
    r"<iframe[^>]*>.*?</iframe>",
    re.IGNORECASE | re.DOTALL,
)
_JAVASCRIPT_URI_RE = re.compile(r"javascript\s*:", re.IGNORECASE)

# Path traversal / SQL-ish fragments in tool args
_DANGEROUS_FRAGMENTS = [
    re.compile(r"\.\./"),
    re.compile(r"';?\s*DROP\s+TABLE", re.IGNORECASE),
    re.compile(r"';?\s*DELETE\s+FROM", re.IGNORECASE),
]


def normalize_unicode(text: str) -> str:
    return unicodedata.normalize("NFC", text)


def strip_control_characters(text: str) -> str:
    if not text:
        return text
    cleaned = _NULL_BYTE_RE.sub("", text)
    return _CONTROL_CHAR_RE.sub("", cleaned)


def remove_active_content(text: str) -> str:
    """Remove script/iframe tags and javascript: URIs from model or tool output."""
    if not text:
        return text
    result = _SCRIPT_TAG_RE.sub("", text)
    result = _IFRAME_TAG_RE.sub("", result)
    result = _JAVASCRIPT_URI_RE.sub("", result)
    return result


def contains_dangerous_fragments(text: str) -> bool:
    if not text:
        return False
    return any(p.search(text) for p in _DANGEROUS_FRAGMENTS)


def sanitize_user_input(text: str, max_length: int = 2000) -> str:
    """
  Prepare user message for the agent: normalize, strip controls, cap length.
  Does not redact PII (booking flows need email/phone until persisted securely).
    """
    if not text or not isinstance(text, str):
        return ""

    cleaned = normalize_unicode(strip_control_characters(text.strip()))
    if len(cleaned) > max_length:
        cleaned = cleaned[:max_length]
    return cleaned


def sanitize_output_text(text: str) -> str:
    """Full pipeline for agent responses: strip active content + redact PII."""
    if not text:
        return text
    return redact_pii(remove_active_content(text))


def sanitize_for_logging(value: Any, max_str_len: int = 500) -> Any:
    """Recursively redact PII and truncate strings before writing logs."""
    if value is None or isinstance(value, (bool, int, float)):
        return value

    if isinstance(value, str):
        truncated = value if len(value) <= max_str_len else f"{value[:max_str_len]}…"
        return redact_pii(truncated)

    if isinstance(value, dict):
        return {str(k): sanitize_for_logging(v, max_str_len) for k, v in value.items()}

    if isinstance(value, (list, tuple)):
        return [sanitize_for_logging(item, max_str_len) for item in value]

    return redact_pii(str(value)[:max_str_len])

File: vinwonders-agent/lib/security/sanitization.ts
/**
 * Input/output sanitization for chat API and structured logs.
 */

import { redactPii } from './redaction';

const CONTROL_CHARS = /[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]/g;
const NULL_BYTES = /\x00/g;
const SCRIPT_TAG = /<script[^>]*>[\s\S]*?<\/script>/gi;
const IFRAME_TAG = /<iframe[^>]*>[\s\S]*?<\/iframe>/gi;
const JS_URI = /javascript\s*:/gi;

const DANGEROUS_FRAGMENTS = [
  /\.\.\//,
  /';?\s*DROP\s+TABLE/i,
  /';?\s*DELETE\s+FROM/i,
];

export function stripControlCharacters(text: string): string {
  return text.replace(NULL_BYTES, '').replace(CONTROL_CHARS, '');
}

export function removeActiveContent(text: string): string {
  return text
    .replace(SCRIPT_TAG, '')
    .replace(IFRAME_TAG, '')
    .replace(JS_URI, '');
}

export function containsDangerousFragments(text: string): boolean {
  return DANGEROUS_FRAGMENTS.some((p) => p.test(text));
}

export function sanitizeUserInput(text: string, maxLength = 2000): string {
  if (!text) return '';
  let cleaned = stripControlCharacters(text.trim());
  if (cleaned.length > maxLength) {
    cleaned = cleaned.slice(0, maxLength);
  }
  return cleaned;
}

export function sanitizeOutputText(text: string): string {
  return redactPii(removeActiveContent(text));
}

export function sanitizeForLogging(
  value: unknown,
  maxStrLen = 500,
): unknown {
  if (value === null || value === undefined) return value;
  if (typeof value === 'boolean' || typeof value === 'number') return value;

  if (typeof value === 'string') {
    const truncated =
      value.length <= maxStrLen ? value : `${value.slice(0, maxStrLen)}…`;
    return redactPii(truncated);
  }

  if (Array.isArray(value)) {
    return value.map((item) => sanitizeForLogging(item, maxStrLen));
  }

  if (typeof value === 'object') {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      out[k] = sanitizeForLogging(v, maxStrLen);
    }
    return out;
  }

  return redactPii(String(value).slice(0, maxStrLen));
}

export function sanitizeToolOutputJson(output: unknown): string {
  const raw = typeof output === 'string' ? output : JSON.stringify(output);
  return sanitizeOutputText(raw);
}
